script/tool/lib/src/gradle_check_command.dart

```diff
@@ -36,31 +36,169 @@ class GradleCheckCommand extends PackageLoopingCommand {
-
-      if (!lines.any((String line) =>
-              line.contains('languageVersion') &&
-              !line.trim().startsWith('//')) &&
-          !lines.any((String line) =>
-              line.contains('sourceCompatibility') &&
-              !line.trim().startsWith('//'))) {
-        const String errorMessage = '''
+    final Directory androidDir =
+        package.platformDirectory(FlutterPlatform.android);
+    final File topLevelGradleFile = _getBuildGradleFile(androidDir);
+
+    // This is tracked as a variable rather than a sequence of &&s so that all
+    // failures are reported at once, not just the first one.
+    bool succeeded = true;
+    if (isExample) {
+      if (!_validateExampleTopLevelBuildGradle(package, topLevelGradleFile)) {
+        succeeded = false;
+      }
+
+      final File appGradleFile =
+          _getBuildGradleFile(androidDir.childDirectory('app'));
+      if (!_validateExampleAppBuildGradle(package, appGradleFile)) {
+        succeeded = false;
+      }
+    } else {
+      succeeded = _validatePluginBuildGradle(package, topLevelGradleFile);
+    }
+
+    return succeeded;
+  }
+
+  // Returns the gradle file in the given directory.
+  File _getBuildGradleFile(Directory dir) => dir.childFile('build.gradle');
+
+  // Returns the main/AndroidManifest.xml file for the given package.
+  File _getMainAndroidManifest(RepositoryPackage package,
+      {required bool isExample}) {
+    final Directory androidDir =
+        package.platformDirectory(FlutterPlatform.android);
+    final Directory baseDir =
+        isExample ? androidDir.childDirectory('app') : androidDir;
+    return baseDir
+        .childDirectory('src')
+        .childDirectory('main')
+        .childFile('AndroidManifest.xml');
+  }
+
+  /// Validates the build.gradle file for a plugin
+  /// (some_plugin/android/build.gradle).
+  bool _validatePluginBuildGradle(RepositoryPackage package, File gradleFile) {
+    print('${indentation}Validating '
+        '${getRelativePosixPath(gradleFile, from: package.directory)}.');
+    final String contents = gradleFile.readAsStringSync();
+    final List<String> lines = contents.split('\n');
+
+    // This is tracked as a variable rather than a sequence of &&s so that all
+    // failures are reported at once, not just the first one.
+    bool succeeded = true;
+    if (!_validateNamespace(package, contents, isExample: false)) {
+      succeeded = false;
+    }
+    if (!_validateSourceCompatibilityVersion(lines)) {
+      succeeded = false;
+    }
+    return succeeded;
+  }
+
+  /// Validates the top-level build.gradle for an example app (e.g.,
+  /// some_package/example/android/build.gradle).
+  bool _validateExampleTopLevelBuildGradle(
+      RepositoryPackage package, File gradleFile) {
+    print('${indentation}Validating '
+        '${getRelativePosixPath(gradleFile, from: package.directory)}.');
+    // TODO(stuartmorgan): Move the -Xlint validation from lint_android_command
```

Comment:
I have a follow-up PR to address this TODO once this one lands; I started to do it as part of the refactoring, but it added a lot of changes that weren't related to the purpose of this PR so I split it out.